libsolidity/interface/FileReader.h

```diff
@@ -0,0 +1,87 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <boost/filesystem.hpp>
+#include <map>
+#include <vector>
+
+#include <libsolidity/interface/ReadFile.h>
+
+namespace solidity {
+
+class FileReader
+{
+public:
+	using StringMap = std::map<std::string, std::string>;
+
+	FileReader(
+		boost::filesystem::path _basePath,
+		std::vector<boost::filesystem::path> _allowedDirectories
+	):
+		m_basePath(std::move(_basePath)),
+		m_allowedDirectories(std::move(_allowedDirectories)),
+		m_sourceCodes(),
+		m_fullPathMapping()
+	{}
+
+	FileReader(): FileReader({}, {}) {}
+
+	boost::filesystem::path const& basePath() const noexcept { return m_basePath; }
+	void setBasePath(boost::filesystem::path const& _path);
+
+	void allowDirectory(boost::filesystem::path _path);
+
+	std::vector<boost::filesystem::path> const& allowedDirectories() const noexcept { return m_allowedDirectories; }
+
+	StringMap const& sourceCodes() const noexcept { return m_sourceCodes; }
+	StringMap& sourceCodes() noexcept { return m_sourceCodes; }
+
+	std::vector<std::string> sourceNames() const;
+	std::string const& sourceCode(std::string const& _name) const { return m_sourceCodes.at(_name); }
+
+	StringMap const& fullPathMapping() const noexcept { return m_fullPathMapping; }
+
+	/// Adds given source, imported as @p _path with filesystem path @p _fspath.
+	void setSource(std::string _path, std::optional<boost::filesystem::path> _fspath, std::string _source);
+
+	/// Resets all sources to the given ones.
+	void setSources(StringMap _sources);
+
+	frontend::ReadCallback::Result readFile(std::string const& _kind, std::string const& _path);
+
+	frontend::ReadCallback::Result operator()(std::string const& _kind, std::string const& _path)
+	{
+		return readFile(_kind, _path);
+	}
+
+private:
+	/// Base path, used for resolving relative paths in imports.
+	boost::filesystem::path m_basePath;
+
+	/// list of allowed directories to read files from
+	std::vector<boost::filesystem::path> m_allowedDirectories;
+
+	/// map of input files to source code strings
+	StringMap m_sourceCodes;
+
+	/// map of input file names to full path names suitable for file://-URIs
+	StringMap m_fullPathMapping;
```

Comment:
> Do we really need `m_fullPathMapping`? The concept of "full path" exists only for files that are loaded from the filesystem. 

When compiling code you usually only deal with the so called source unit IDs, and there just having the basic FileReader interface that already did exist was fine, as it was mapping that to the local disk if needed to fetch the source code.

The language client however passes file URIs to the language server, and I need to be able map these back to the source unit ID (and vice versa). But maybe there is a better way?

[...]
> Moreover the file on disk can change after we have loaded it so it does not seem safe to refer to the file using this path.

I think this could have happened before either, and we didn't care about because `solc` usually is short-living. However, when running in language server mode, `solc` is a long running process and then it might make sense to consider this.
However, the client is telling us about any changes the user did (within the client IDE, - not from the file system).

Reply:
> I think this could have happened before either, and we didn't care about because solc usually is short-living. 

AFAIK it could not. We load a file into the virtual filesystem once and then `CompilerStack` uses that whenever the file is referenced. If you manage to make it load the file again it's only because it thinks it's a different file (i.e. it has a different source unit ID) and it will be compiled again and treated as a separate source unit.

> The language client however passes file URIs to the language server, and I need to be able map these back to the source unit ID (and vice versa). But maybe there is a better way?

How does the language server access files from Standard JSON or compiler's standard input? Or from JSON passed to the standard input? :)

Also, do you support files loaded from disk like this?
```json
       {
           "language": "Solidity",
           "sources": {
               "/usr/local/dapp-bin/token.sol": {
                   "urls": ["/projects/mytoken.sol"]
               }
           },
           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
       }
```

I think it would be best if there was a way for the compiler to share the source it has already loaded into its virtual filesystem.

If that is not possible and the LSP is meant to work only with files that were loaded from disk then I think you could just replace this map with a function that calculates the path based on source unit ID or the other way. That should be deterministic as long as you know that it was not loaded from some other source (we might have to add some enum to indicate where it came from).